### ISSUE TITLE
Tracer error with eqx.jit, jax.grad and dfx.diffeqsolve

### ISSUE THREAD

**Issue by jatentaki** [patrick-kidger/equinox#99]
I'm experiencing an exception in a code which tries to backpropagate through `diffeqsolve`, the whole thing being wrapped in `eqx.filter_jit`. The context is essentially an extension of your [score-based diffusion](https://docs.kidger.site/equinox/examples/score_based_diffusion/) example to calculate likelihoods as shown [here](https://colab.research.google.com/drive/1SeXMpILhkJPjXUaesvzEhc3Ke6Zl_zxJ?usp=sharing). This works correctly, however when I try to take `jax.grad` of the log-likelihood w.r.t. sample (image), I get an error about a leaked tracer; indeed the leak location does look like it stores something in a struct. The reproducing code along with the error message are [here](https://colab.research.google.com/drive/1OactTAggitMIhlBA-LHltf-GxUjk4BBF?usp=sharing). I hope this is not too difficult to understand, it's mostly refactored code of yours :)

**Comment by patrick-kidger** [comment]
Unfortunately the repro is far too large for me to be able to dig into! (Even if it is derived from some of my code.) There's a lot that can be removed here to make it smaller and more readable: the `dimchecked` annotations, the training etc. happening in the middle; the iterations using `tqdm`; lines like `pt = pt - 0.01 * grad` which whilst mathematically appropriate probably aren't the cause of the software error; etc.

(On a side note -- I'm very interested to see a reference to `torch-dimcheck`, which wasn't something I was previously aware of. I wrote [TorchTyping](https://github.com/patrick-kidger/torchtyping), which does very something similar.)

**Comment by jatentaki** [comment]
Sure, I thought the repro may be enough to diagnose what's going wrong. I created a [new version](https://colab.research.google.com/drive/1yjQebChsLWuGNVVVfyp27OHEnd_DEAfh?usp=sharing), can't guarantee it's truly minimal but certainly shorter to read. Would this be enough for you to see what's going on?

**Comment by patrick-kidger** [comment]
Yep, with this version it's immediately clear what the issue is!

When you have `def ode_fun` then `self.eq.weight` is being sneakily included via closure. Most of the time this will work without issue: JAX just sees the closed-over value when the closure is eventually evaluated and life proceeds as normal.

Meanwhile, when you have `ode_func = self.eq` then `terms` is a PyTree containing `self.eq`, which in turn is a PyTree containing `self.eq.weight`: this whole structure is visible to `diffeqsolve` -- and importantly, it is visible to `BacksolveAdjoint`.

The specific purpose of `BacksolveAdjoint` is to override how autodiff is performed, and in particular it will change how gradients are handled with respect to `terms`, `y0`, and `args`. When `terms` is a PyTree then it can see `self.eq.weights`, and there are no "implicit" inputs into the calculation. But when the PyTree structure is broken then we do have such an "implicit" input sneaking in to the calculation -- and we get the tracer error that you then see.

Various comments:

1. A related error to this is stated in the FAQ [here](https://docs.kidger.site/diffrax/further_details/faq/#im-getting-zero-gradient-for-one-of-my-model-parameters) but (a) this is obviously a slightly different issue, and (b) I don't think it's actually stated in the `BacksolveAdjoint` docs, which should probably be fixed. (Open a PR?)
2. Fixing this "once and for all" (in a robust hard-to-accidentally-write-errors way) will require changes to how `custom_vjp` works in core JAX.
3. At the moment you're doing `grad-jit`. I'd recommend doing `jit-grad` instead. A general principle for using JAX is that it's worth having a single JIT at the top level, wrapping everything else. (And in this particular case, it will actually also avoid raising an error -- we don't get an escaped tracer error because the JIT is outside the custom gradient handling, but conversely because `self.eq.weight` isn't visible to the custom gradient handling then if you were to try and compute a gradient with it then you'd get a zero gradient.)

**Comment by jatentaki** [comment]
Great, thanks :)
